File: 프로그래머스/3/87694. 아이템 줍기/아이템 줍기.js
// Magnify the grid twice to find all the edge cases
// Consider some curved cases
// Check if any of the 8-direction points are not occupied by the shape
// Run dfs(or bfs) to find the shortest path. bfs should be okay since in-degre of every node in graph <= 1
// Time complexity: board length : n. O(n^2)
// Space complextiy: O(n^2)

class Queue {
    constructor() {
        this.front = [];
        this.back = [];
    }
    
    get size() {
        return this.front.length + this.back.length;
    }
    
    get isEmpty() {
        return this.size === 0;
    }
    
    enqueue(element) {
        this.back.push(element);
    }
    
    dequeue() {
        if (this.isEmpty) return null;
        
        if (this.front.length) {
            return this.front.pop();
        }
        
        while (this.back.length) {
            this.front.push(this.back.pop());
        }
        
        return this.front.pop();
    }
}

const getAllAroundPositions = ([ x, y ]) => [
    [ x - 1, y ],
    [ x + 1, y ],
    [ x, y - 1],
    [ x, y + 1 ],
    [ x - 1, y - 1],
    [ x - 1, y + 1 ],
    [ x + 1, y - 1],
    [ x + 1, y + 1]
];

const getNearbyPositions = ([ x, y ]) => [
    [ x - 1, y ],
    [ x + 1, y ],
    [ x, y - 1 ],
    [ x, y + 1 ],
]

const checkValidPosition = (board, [ x, y ]) => x >= 0 && x < board.length && y >= 0 && y < board[0].length;

const getValidAllAroundPositions = (board, position) => getAllAroundPositions(position).filter((position) => checkValidPosition(board, position));

const getValidNearbyPositions = (board, position) => getNearbyPositions(position).filter((position) => checkValidPosition(board, position));

const Json = (value) => JSON.stringify(value);

const checkPositionEqual = ([ x1, y1 ], [ x2, y2 ]) => x1 === x2 && y1 === y2;

function solution(rectangles, characterX, characterY, itemX, itemY) {
    const board = Array.from({ length: 102, }, () => Array.from({ length: 102 }, () => 0));
    
    rectangles.forEach(([ x1, y1, x2, y2 ]) => {
        for (let i = 2 * Math.min(x1, x2); i <= 2 * Math.max(x1, x2); i++) {
            for (let j = 2 * Math.min(y1, y2); j <= 2 * Math.max(y1, y2); j++) {
                board[i][j] = 1;
            }
        }
    });
    
    for (let i = 0; i < board.length; i++) {
        for (let j = 0; j <= board[0].length; j++) {
            if (board[i][j] === 0) continue;
            const positions = getValidAllAroundPositions(board, [ i, j ]);
            if (!positions.some(([ x, y ]) => board[x][y] === 0)) continue;
            
            board[i][j] = 2;
        }
    }
    
    const queue = new Queue();
    const source = [ characterX * 2, characterY * 2 ]
    const root = {
        node: source,
        distance: 0,
    }
    const destination = [ itemX * 2, itemY * 2 ];
    queue.enqueue(root);
    
    let shortestDistance = Infinity;
    const visited = new Set();
    visited.add(Json(source));
    
    while (!queue.isEmpty) {
        const { node, distance } = queue.dequeue();
        const [ x, y ] = node;
        
        if (checkPositionEqual(node, destination)) {
            shortestDistance = Math.min(shortestDistance, distance);
            return shortestDistance / 2;
        }
        
        const positions = getValidNearbyPositions(board, node).filter(([ x, y ]) => board[x][y] === 2 && !visited.has(Json([ x, y ])));
        
        
        positions.forEach(([ neighborX, neighborY ]) => {
            queue.enqueue({
                node: [ neighborX, neighborY ],
                distance: distance + 1,
            });
            visited.add(Json([ neighborX, neighborY ]));
        });
    }
}

File: 프로그래머스/3/87694. 아이템 줍기/아이템 줍기.py
from collections import deque

def solution(rectangles, characterX, characterY, itemX, itemY):
    matrix = [ [ 0 for j in range(0, 200) ] for i in range(0, 200) ]
    
    for rectangle in rectangles:
        x1, y1, x2, y2 = [ number * 2 for number in rectangle]
        for x in range(x1, x2 + 1):
            for y in range(y1, y2 + 1):
                matrix[x][y] = 1
            
    root = (2 * characterX, 2 * characterY,)
    target = (2 * itemX, 2 * itemY,)

    return bfs(matrix, root, target, set()) / 2
            
                
    
def bfs(matrix, root, target, visited):
    queue = deque([ (root, 0,) ])
    visited.add(root)
    
    while len(queue) > 0:
        current, distance = queue.pop()
        
        x, y = current
        
        if target == current:
            return distance
            
        corners = [ (x - 1, y), (x + 1, y), (x, y - 1), (x, y + 1 ), (x - 1, y - 1), (x - 1, y + 1), (x + 1, y - 1), (x + 1, y + 1) ]
        
        isAllFilled = all([ matrix[corner_x][corner_y] == 1 for corner_x, corner_y in corners ])
        
        if isAllFilled:    
            continue
            
        next_nodes = [ (x - 1, y), (x + 1, y), (x, y - 1), (x, y + 1 ) ]

        for next_node in next_nodes:
            if next_node in visited:
                continue
                
            next_x, next_y = next_node
            if matrix[next_x][next_y] == 0:
                continue
                
            visited.add(next_node)
            queue.appendleft((next_node, distance + 1,))

            
    return 10000
            
        
        
    
    
    
#     min_table_x = {}
#     max_table_x = {}
#     min_table_y = {}
#     max_table_y = {}
    
#     for rectangle in rectangles:
#         x1, y1, x2, y2 = rectangle
#         # 1. x1 그대로, y1 => y2
#         # 2. x2 그대로, y1 => y2
#         for y in range(y1, y2 + 1):
#             if y not in min_table_y:
#                 min_table_y[y] = x1
#                 max_table_y[y] = x2
                
#             min_table_y[y] = min(min_table_y[y], x1)
#             max_table_y[y] = max(min_table_y[y], x2)
            
#         for x in range(x1, x2 + 1):
#             if x not in min_table_x:
#                 min_table_x[x] = y1
#                 max_table_x[x] = y2
                
#             min_table_x[x] = min(min_table_x[x], y1)
#             max_table_x[x] = max(min_table_x[x], y2)
            
#     valid_positions = set()
    
#     for x in min_table_x:
#         y = min_table_x[x]
#         valid_positions.add((x, y,))
    
#     for x in min_table_x:
#         y = max_table_x[x]
#         valid_positions.add((x, y,))
        
#     for y in min_table_y:
#         x = min_table_y[y]
#         valid_positions.add((x, y,))
        
#     for y in max_table_y:
#         x = max_table_y[y]
#         valid_positions.add((x, y,))
        
#     print((4,5,) in valid_positions)
        
        
#     print(valid_positions)
        
        # 3. y1 그대로, x1 => x2
        # 4. y2 그대로, x1 => x2
        
    
    
    answer = 0
    return answer

# 해결책 1
# 각 x좌표에서 최소값/최대값, 각 y좌표에서 최소값/최대값 구함
# 모여서 한 다각형을 이루므로 이 좌표들만 이동할 수 있는 유효한 값
# bfs로 다음 노드 이동할 때 해당 좌표들로만 이동 가능
# 문제: 최대값 최소값이 아닌 중간 값중에도 유효한 값이 있는데 이를 유효하지 않은 값으로 간주.

# 해결책 2
# 주어지는 다각형 영역을 1로 표시
# 좌표계 확장: 2배로 확장.(좌표계 확장)
# bfs 탐색을 진행하되, 
# 다각형 영역에 포함되는 부분만 탐색, 
# 그리고 주변(상하좌우) 좌표가 모두 다각형에 포함되는 경우 버림
# 거리 = 확장된 좌표계에서의 거리 / 2




# 최대 사이즈 100 * 100 = 10000
# 1. 지형 테두리를 모두 칠한다
# 2. 가장자리에서 어느 방향으로 가야 할까?
#    